File: src/core/execution/nodes/telegram_bot.py
"""
Telegram Bot Node
Sends messages to Telegram groups/channels via bot API
"""
from typing import Dict, Any, Optional
from ..node_base import BaseNode, ExecutionContext
from src.utils.logger import get_logger

logger = get_logger(__name__)


class TelegramBotNode(BaseNode):
    """
    Sends messages to Telegram chats via bot API
    
    Inputs:
        message: Message text to send (required)
        bot_id: Telegram bot token (optional, can be provided as widget or input)
        chat_id: Telegram chat ID (optional, can be provided as widget or input)
                 Also accepts group_id for backwards compatibility
    
    Outputs:
        success: Boolean indicating if message was sent successfully
        response: Response from Telegram API (if available)
    """
    
    def __init__(self, node_id: str, node_data: Dict[str, Any]):
        """Initialize Telegram bot node"""
        super().__init__(node_id, node_data)
    
    def execute(self, context: ExecutionContext) -> Dict[str, Any]:
        """Execute Telegram bot node - send message to Telegram"""
        # Get message (required, should come from input)
        message = self.get_input_value('message', context, None)
        
        # Get bot_id: try input first, then metadata (widget value)
        bot_id = self.get_input_value('bot_id', context, None)
        if not bot_id:
            bot_id = self.metadata.get('bot_id', '') or self.inputs.get('bot_id', '')
        
        # Get chat_id: try chat_id first, then group_id for backwards compatibility
        chat_id = self.get_input_value('chat_id', context, None)
        if not chat_id:
            chat_id = self.get_input_value('group_id', context, None)  # backwards compat
        if not chat_id:
            chat_id = self.metadata.get('chat_id', '') or self.metadata.get('group_id', '') or self.inputs.get('chat_id', '') or self.inputs.get('group_id', '')
        
        # Resolve template variables (including environment variables)
        import os
        
        def resolve_template_var(value: str) -> str:
            """Resolve template variable, including environment variables"""
            if isinstance(value, str) and value.startswith('{{') and value.endswith('}}'):
                var_name = value[2:-2].strip()
                # Check if it's an environment variable (process.env.VAR_NAME)
                if var_name.startswith('process.env.'):
                    env_var = var_name.replace('process.env.', '')
                    return os.getenv(env_var, None)
                else:
                    return context.variables.get(var_name, None)
            return value
        
        message = resolve_template_var(message) if message else None
        bot_id = resolve_template_var(bot_id) if bot_id else None
        chat_id = resolve_template_var(chat_id) if chat_id else None
        
        # Validate inputs
        if not message:
            raise ValueError("message is required for TelegramBotNode")
        
        if not bot_id:
            raise ValueError("bot_id is required for TelegramBotNode")
        
        if not chat_id:
            raise ValueError("chat_id is required for TelegramBotNode")
        
        try:
            # Import requests for API calls
            import requests
            
            # Telegram Bot API endpoint
            url = f"https://api.telegram.org/bot{bot_id}/sendMessage"
            
            # Prepare payload
            payload = {
                "chat_id": chat_id,
                "text": str(message),
                "parse_mode": "HTML"  # Optional: supports HTML formatting
            }
            
            # Send message
            logger.debug(f"[TelegramBot] Sending message to chat_id={chat_id}, message_length={len(str(message))}")
            response = requests.post(url, json=payload, timeout=10)
            response.raise_for_status()
            
            result = response.json()
            
            if result.get('ok'):
                logger.info(f"[TelegramBot] Message sent successfully to chat_id={chat_id}")
                return {
                    'success': True,
                    'response': result
                }
            else:
                error_msg = result.get('description', 'Unknown error')
                logger.error(f"[TelegramBot] Failed to send message: {error_msg}")
                return {
                    'success': False,
                    'response': result
                }
                
        except requests.exceptions.RequestException as e:
            logger.error(f"[TelegramBot] Request error: {e}")
            return {
                'success': False,
                'response': {'error': str(e)}
            }
        except Exception as e:
            logger.error(f"[TelegramBot] Unexpected error: {e}")
            return {
                'success': False,
                'response': {'error': str(e)}
            }

File: ts/src/core/execution/nodes/telegramBot.ts
/**
 * TelegramBotNode – sends a message to Telegram via the Bot API.
 * Mirrors Python src/core/execution/nodes/telegram_bot.py
 */
import { BaseNode, ExecutionContext } from "../nodeBase";
import { getLogger } from "../../../utils/logger";

const logger = getLogger("telegramBot");

export class TelegramBotNode extends BaseNode {
  async execute(context: ExecutionContext): Promise<Record<string, unknown>> {
    const message = this.getInputValue("message", context, undefined) as
      | string
      | undefined;
    const botToken =
      (this.getInputValue("bot_id", context, undefined) as string) ||
      (this.getInputValue("bot_token", context, undefined) as string) ||
      (this.resolveEnvVar(this.metadata.bot_id) as string) ||
      (this.resolveEnvVar(this.metadata.bot_token) as string) ||
      process.env.TELEGRAM_DEV_AGENT_BOT_TOKEN ||
      process.env.TELEGRAM_BOT_TOKEN ||
      "";
    const chatId =
      (this.getInputValue("chat_id", context, undefined) as string) ||
      (this.resolveEnvVar(this.metadata.chat_id) as string) ||
      process.env.TELEGRAM_CHAT_ID ||
      "";

    // Gracefully handle gated messages (e.g. BinaryIntentNode returned null)
    if (!message) {
      logger.debug(
        `[TelegramBot] No message provided (likely gated by BinaryIntent), skipping send for node ${this.nodeId}`
      );
      return {
        success: false,
        response: { error: "No message provided, skipped." },
      };
    }

    if (!botToken) {
      throw new Error(
        `TelegramBotNode ${this.nodeId}: bot_token is required`
      );
    }
    if (!chatId) {
      throw new Error(
        `TelegramBotNode ${this.nodeId}: chat_id is required`
      );
    }

    // Optional message_id for quote-replying
    const replyToMessageId =
      (this.getInputValue("message_id", context, undefined) as number | string | undefined);
    const replyId = replyToMessageId ? Number(replyToMessageId) : undefined;

    const url = `https://api.telegram.org/bot${botToken}/sendMessage`;

    // Build payload — include reply_to_message_id when available
    const payload: Record<string, unknown> = {
      chat_id: chatId,
      text: message,
      parse_mode: "HTML",
    };
    if (replyId && Number.isFinite(replyId)) {
      payload.reply_parameters = { message_id: replyId };
    }

    try {
      let res = await fetch(url, {
        method: "POST",
        headers: { "Content-Type": "application/json" },
        body: JSON.stringify(payload),
      });

      let data = (await res.json()) as Record<string, unknown>;

      // If quote-reply failed (e.g. message deleted), retry without it
      if ((!res.ok || !data.ok) && payload.reply_parameters) {
        logger.warn(
          `[TelegramBot] Quote-reply failed (message_id=${replyId}), retrying without reply: ${(data as any).description ?? ""}`
        );
        delete payload.reply_parameters;
        res = await fetch(url, {
          method: "POST",
          headers: { "Content-Type": "application/json" },
          body: JSON.stringify(payload),
        });
        data = (await res.json()) as Record<string, unknown>;
      }

      if (!res.ok || !data.ok) {
        logger.error(
          `[TelegramBot] API error: ${JSON.stringify(data).slice(0, 300)}`
        );
        return { success: false, response: data };
      }

      logger.debug(
        `[TelegramBot] Message sent to chat ${chatId} (${message.length} chars)${replyId ? ` reply_to=${replyId}` : ""}`
      );
      return { success: true, response: data };
    } catch (err) {
      const msg = err instanceof Error ? err.message : String(err);
      logger.error(`[TelegramBot] Failed to send: ${msg}`);
      return { success: false, response: { error: msg } };
    }
  }
}
